### PR TITLE
[WIP] Error API

### PR DESCRIPTION
**Pull request by peter-edge**
Not ready for review, but for reference. I'll split this up into multiple PRs when ready for review.

### CODE REVIEW

File: api/errors/errors.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+// TODO: What to do with ERROR_TYPE_UNKNOWN? Should we use use this for ERROR_TYPE_APPLICATION?
+// TODO: What to do with ERROR_TYPE_INTERNAL? Should we have a function to create an error of this type?
+// TODO: should we have specific fields for each error type instead of keyValues ...string?
+
+// Type will extract the ErrorType from a yarpc error.
+//
+// This will return ERROR_TYPE_NONE if the given error is nil or not a yarpc error.
+// This function is the defined way to test if an error is a yarpc error.
+func Type(err error) yarpcproto.ErrorType {
```
Comment (willhug) [inline api/errors/errors.go]: probably shouldn't be proto right?

File: yarpcproto/yarpc.proto [hunk 2]
```diff
@@ -3,9 +3,190 @@ syntax = "proto3";
...
+//
+// Sometimes multiple error codes may apply. Services should return
+// the most specific error code that applies. For example, prefer
+// `OUT_OF_RANGE` over `FAILED_PRECONDITION` if both codes apply.
+// Similarly prefer `NOT_FOUND` or `ALREADY_EXISTS` over `FAILED_PRECONDITION`.
+enum ErrorType {
+  // Not an error; returned on success
+  //
+  // HTTP Mapping: 200 OK
+  // Google API Mapping: Code.OK
```
Comment (willhug) [inline yarpcproto/yarpc.proto]: s/Google/GRPC/ ?

File: api/errors/errors.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+	return copyStringStringMap(yarpcError.Details)
+}
+
+// WithKeyValues adds the given keyValues to the error.
+//
+// If the error is a yarpc error, this will just add the keyValues.
+// If the error is not a yarpc error, this will return a new yarpc error of type ERROR_TYPE_UNKNOWN
+// with an additional keyValue pair of "error", err.Error().
+// If keyValues contains a key of "error" and a new yarpc error is created, this key will be overwritten.
+func WithKeyValues(err error, keyValues ...string) error {
```
Comment (willhug) [inline api/errors/errors.go]: wiw9w

File: yarpcproto/yarpc.proto (hunk 2 above)
Comment (peter-edge) [inline yarpcproto/yarpc.proto]: Its actually google in this case https://godoc.org/google.golang.org/genproto/googleapis/rpc/code

File: api/errors/errors.go (hunk 1 above)
Comment (peter-edge) [inline api/errors/errors.go]: Ya I'm moving this per a convo with Kris and Abhinav

**Pull request closed without merging**